### DISCUSSION TITLE
get_examples function in Tok2Vec.initialize [E930]

### DISCUSSION THREAD

**Discussion by m-nlp-q** [explosion/spaCy#7273] [Help: Coding & Implementations] [answered]
I am constructing the training `config` file and when I run the `debug data` command I get the following error: 

`Received invalid get_examples callback in `Tok2Vec.initialize`. Expected function that returns an iterable of Example objects but got: []`

It seems that a function that should be placed in `[initialize.tok2vec]` and returns an iterable of Example is missing. Should I write my own function or can I reference a function provided by spaCy?

My task is NER and the data is in IOB2 format. I have converted my data into `.spacy` format.

**Comment by adrianeboyd** [comment, marked as answer]
Hmm, the data needed for `Tok2Vec.initialize` should come from the train corpus specified in the config, but it's a bit of catch-22 if you can't debug the data because you can't initialize the pipeline.

Let's see, can you first check that the config is okay with `spacy debug config`?

**Reply by m-nlp-q** [reply]
@adrianeboyd thank you for your reply.

I had already checked with `spacy debug config`, which runs successfully. `spacy train` led me to `spacy debug data`, because when I run `spacy train` it throws the error:

`It looks like there is no proper sample data to initialize the Model of component 'tok2vec'. To check your input data paths and annotation, run: python -m spacy debug data config.cfg`

This is the output of the `spacy debug config`: 

```
2021-03-05 07:27:43.201478: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0

============================= Config validation =============================

===================== Config validation for [initialize] =====================

====================== Config validation for [training] ======================
✔ Config is valid

=============================== Variables (0) ===============================

Variable                                   Value                             
-----------------------------------------  ----------------------------------


========================= Registered functions (16) =========================
ℹ [components.ner.model]
Registry   @architectures                
Name       spacy.TransitionBasedParser.v2
Module     spacy.ml.models.parser        
File       /usr/local/lib/python3.7/dist-packages/spacy/ml/models/parser.py (line 34)
ℹ [components.ner.model.tok2vec]
Registry   @architectures                
Name       spacy-legacy.Tok2Vec.v1       
Module     spacy_legacy.architectures.tok2vec
File       /usr/local/lib/python3.7/dist-packages/spacy_legacy/architectures/tok2vec.py (line 8)
ℹ [components.ner.model.tok2vec.embed]
Registry   @architectures                
Name       spacy.MultiHashEmbed.v1       
Module     spacy.ml.models.tok2vec       
File       /usr/local/lib/python3.7/dist-packages/spacy/ml/models/tok2vec.py (line 111)
ℹ [components.ner.model.tok2vec.encode]
Registry   @architectures                
Name       spacy-legacy.MaxoutWindowEncoder.v1
Module     spacy_legacy.architectures.tok2vec
File       /usr/local/lib/python3.7/dist-packages/spacy_legacy/architectures/tok2vec.py (line 28)
ℹ [components.tok2vec.model]
Registry   @architectures                
Name       spacy-legacy.Tok2Vec.v1       
Module     spacy_legacy.architectures.tok2vec
File       /usr/local/lib/python3.7/dist-packages/spacy_legacy/architectures/tok2vec.py (line 8)
ℹ [components.tok2vec.model.embed]
Registry   @architectures                
Name       spacy.MultiHashEmbed.v1       
Module     spacy.ml.models.tok2vec       
File       /usr/local/lib/python3.7/dist-packages/spacy/ml/models/tok2vec.py (line 111)
ℹ [components.tok2vec.model.encode]
Registry   @architectures                
Name       spacy-legacy.MaxoutWindowEncoder.v1
Module     spacy_legacy.architectures.tok2vec
File       /usr/local/lib/python3.7/dist-packages/spacy_legacy/architectures/tok2vec.py (line 28)
ℹ [corpora.dev]
Registry   @readers                      
Name       spacy.Corpus.v1               
Module     spacy.training.corpus         
File       /usr/local/lib/python3.7/dist-packages/spacy/training/corpus.py (line 21)
ℹ [corpora.train]
Registry   @readers                      
Name       spacy.Corpus.v1               
Module     spacy.training.corpus         
File       /usr/local/lib/python3.7/dist-packages/spacy/training/corpus.py (line 21)
ℹ [initialize.components.ner.labels]
Registry   @readers                      
Name       spacy.read_labels.v1          
Module     spacy.training.corpus         
File       /usr/local/lib/python3.7/dist-packages/spacy/training/corpus.py (line 48)
ℹ [initialize.lookups]
Registry   @misc                         
Name       spacy.LookupsDataLoader.v1    
Module     spacy.language                
File       /usr/local/lib/python3.7/dist-packages/spacy/language.py (line 92)
ℹ [nlp.tokenizer]
Registry   @tokenizers                   
Name       spacy.Tokenizer.v1            
Module     spacy.language                
File       /usr/local/lib/python3.7/dist-packages/spacy/language.py (line 66)
ℹ [training.batcher]
Registry   @batchers                     
Name       spacy.batch_by_words.v1       
Module     spacy.training.batchers       
File       /usr/local/lib/python3.7/dist-packages/spacy/training/batchers.py (line 49)
ℹ [training.batcher.size]
Registry   @schedules                    
Name       compounding.v1                
Module     thinc.schedules               
File       /usr/local/lib/python3.7/dist-packages/thinc/schedules.py (line 43)
ℹ [training.logger]
Registry   @loggers                      
Name       spacy.WandbLogger.v1          
Module     spacy.training.loggers        
File       /usr/local/lib/python3.7/dist-packages/spacy/training/loggers.py (line 104)
ℹ [training.optimizer]
Registry   @optimizers                   
Name       Adam.v1                       
Module     thinc.optimizers              
File       /usr/local/lib/python3.7/dist-packages/thinc/optimizers.py (line 58)
```

**Reply by adrianeboyd** [reply]
Then I think it's likely that your training corpus is empty. (This is not a very helpful error message for this case!)

Try this to look at the annotation in your training data: https://github.com/explosion/spaCy/discussions/7281#discussioncomment-430588

**Reply by m-nlp-q** [reply]
@adrianeboyd I tried the your code snip of #7281 and it seems that my training data is in the right format. The output of your code is 

```
IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.
```

That is because I run it on colab. I also checked your code with `dev` dataset and it displays the dataset.

My dataset has the `biluo` format and the data within the `.biluo` file is organised as a word per line separated by a white space from the label. This is a small sample of the file:

```
πως O
η O
Αργεντινή U-LOC
ζήτησε O
από O
το O
ΔΝΤ U-ORG
```

In order to covert my data to `.spacy` format, I used the [convert ](https://spacy.io/api/cli#convert) command. Firstly, I converted biluo to IOB2 format to convert my data like the penultimate data sample presented [here](https://github.com/explosion/spaCy/tree/master/extra/example_data/ner_example_data), and use the `ner` converter. The command run successfully and created the `.spacy` file

2021-02-17 09:18:40.306447: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.1
ℹ Auto-detected token-per-line NER format
⚠ No sentence boundaries found to use with option `-n 1`. Use `-s` to
automatically segment sentences or `-n 0` to disable.
⚠ No sentence boundaries found. Use `-s` to automatically segment
sentences.
⚠ No document delimiters found. Use `-n` to automatically group
sentences into documents.
✔ Generated output file (1 documents):
/content/drive/MyDrive/spaCy_NER/train.spacy

What might be the source of the problem?

b.t.w. Since `biluo` is a better format, can I keep it to further train the model? (rem. the model is `el_core_news_lg`)

p.s. Just to clear this out, normally, I don't have to write a custom `get_examples()` function. From your answer I understand that the source of the error is something else.

**Reply by adrianeboyd** [reply]
The source of the error is probably that something went wrong with the conversion and `train.spacy` doesn't contain any docs.

If you have BILUO data like this without blank lines between sentences, one recommended conversion for Greek would be:

```bash
python -m spacy convert -c ner -b el_core_news_sm -n 10 -s file.iob .
```

This adds sentence boundaries by running `el_core_news_sm` (`-b` and `-s`) on the text and groups every 10 sentences (`-n`) into a document. Typically it's better if you have shorter documents that you can shuffle easily for training rather than one long document.

The `Doc` API only shows IOB2 tags, but the internal NER component always uses BILUO tags underneath (no matter the input format), so you don't need to do anything extra to use BILUO for your corpus.

**Reply by m-nlp-q** [reply]
@adrianeboyd sorry for the late response.

I converted my datasets to `.spacy` with the command you suggested:

```
python -m spacy convert \
    /content/training_configuration/dev.iob \
    /content/training_configuration \
    -c ner \
    -b el_core_news_sm \
    -n 10 \
    -s
```

and get a successful output without warnings 

```
2021-03-16 14:51:56.507241: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
ℹ Auto-detected token-per-line NER format
ℹ Grouping every 10 sentences into a document.
ℹ Segmenting sentences with parser from model 'el_core_news_sm'.
✔ Generated output file (1 documents):
/content/training_configuration/dev.spacy
```

**but I still get the same error [E930]**

I use the config file of `el_core_news_ls` and the only modifications that I have made are: 

1. remove components other than `Tok2Vec` and `ner`
2. change deprecated components to `spacy-legacy`
3. and initialize labels with `spacy init labels` 

This is my config file

[paths]
train = "/content/training_configuration/train.spacy"
dev = "/content/training_configuration/dev.spacy"
vectors = null
init_tok2vec = null
vocab_data = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "el"
pipeline = ["tok2vec","ner"]
batch_size = 256
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy-legacy.Tok2Vec.v1"

[components.ner.model.tok2vec.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = 96
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = true

[components.ner.model.tok2vec.encode]
@architectures = "spacy-legacy.MaxoutWindowEncoder.v1"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy-legacy.Tok2Vec.v1"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.tok2vec.model.encode:width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy-legacy.MaxoutWindowEncoder.v1"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
limit = 0
max_length = 0
path = ${paths:dev}
gold_preproc = false
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths:train}
max_length = 5000
gold_preproc = false
limit = 0
augmenter = null

[training]
train_corpus = "corpora.train"
dev_corpus = "corpora.dev"
seed = ${system:seed}
gpu_allocator = ${system:gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 5000
max_epochs = 0
max_steps = 0
eval_frequency = 1000
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.WandbLogger.v1"
project_name = "spacy-v3.0.0a2"
remove_config_values = []

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = true
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
sents_p = null
sents_r = null
ents_per_type = null
sents_f = 0.0
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0

[initialize]
vocab_data = ${paths.vocab_data}
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
before_init = null
after_init = null

[initialize.components]

[initialize.components.ner]

[initialize.components.ner.labels]
@readers = "spacy.read_labels.v1"
path = "/content/corpus/ner.json"
require = false

[initialize.lookups]
@misc = "spacy.LookupsDataLoader.v1"
lang = ${nlp.lang}
tables = ["lexeme_norm"]

[initialize.tokenizer]

**Reply by m-nlp-q** [reply]
I also tested [this ](https://github.com/explosion/spaCy/discussions/6472#discussioncomment-232263)code and got this output: 

```
TRAIN docs: 1 with 27885 entities
DEV docs: 1 with 2984 entities
```

which probably means that the `.spacy` data are correct.

Any idea what is wrong?

**Reply by adrianeboyd** [reply]
You shouldn't need to use `spacy-legacy` in your config. I don't think it will be invalid, but the `spacy`->`spacy-legacy` conversion/lookup should happen internally automatically.

If you only have 1 document with 27885 entities something has probably gone wrong in the conversion, because it was supposed to break the document into paragraphs of 10 sentences each. Is there any way you can share a bit of your data, maybe at least 100-200 lines or so? You can anonymize the token text (just pick a dummy Greek word, keeping punctuation in place if possible for testing the sentence detection part) as long as the IOB tag column stays in place. If you don't want to post it here, you can email me at adriane AT explosion.ai. It's possible there's a bug in the converter, but I am not really sure at this point.

To start from a clean slate mainly to test the conversion, you can try using this project as a starting point: https://github.com/explosion/projects/tree/v3/pipelines/ner_demo

```bash
spacy project clone pipelines/ner_demo
```

Replace the provided convert command in `project.yml` with something like this (using the train/dev variables if you'd like, this is the just supposed to show the commands):

```yaml
  - name: "convert"
    help: "Convert the data to spaCy's binary format"
    script:
      - "python -m spacy convert -c ner -b el_core_news_sm -n 10 -s assets/train.iob corpus/"
      - "python -m spacy convert -c ner -b el_core_news_sm -n 10 -s assets/dev.iob corpus/"
    deps:
      - "assets/train.iob"
      - "assets/dev.iob"
    outputs:
      - "corpus/train.spacy"
      - "corpus/dev.spacy"
```

The project generates a reasonable default config as one of its steps.

**Reply by adrianeboyd** [reply]
Thanks for sharing your data, it was helpful to be able to see exactly what was going on. There is indeed a bug in the converter when it tries to use the parser from the model to split sentences.

If you switch to the `sentencizer` instead as a workaround (leave out `-b el_core_news_sm`), the following steps worked for me:

```bash
python -m spacy convert -n 10 -s train.iob .
python -m spacy convert -n 10 -s dev.iob .
python -m spacy init config -p ner -l el config.cfg
python -m spacy train config.cfg --paths.train ./train.spacy --paths.dev ./dev.spacy
```

Output:

```none
=========================== Initializing pipeline ===========================
[2021-03-18 11:20:22,902] [INFO] Set up nlp object from config
[2021-03-18 11:20:22,912] [INFO] Pipeline: ['tok2vec', 'ner']
[2021-03-18 11:20:22,915] [INFO] Created vocabulary
[2021-03-18 11:20:22,915] [INFO] Finished initializing nlp object
[2021-03-18 11:20:29,442] [INFO] Initialized pipeline components: ['tok2vec', 'ner']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['tok2vec', 'ner']
ℹ Initial learn rate: 0.001
E    #       LOSS TOK2VEC  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  ------------  --------  ------  ------  ------  ------
  0       0          0.00    167.39    0.00    0.00    0.00    0.00
  0     200       5951.27  10171.19   57.21   58.81   55.70    0.57
  0     400        292.78   3506.98   67.65   68.92   66.42    0.68
```

`python -m spacy debug data config.cfg ./train.spacy --paths.dev ./dev.spacy --verbose` shows:

```none
========================== Named Entity Recognition ==========================
ℹ 0 new label(s), 4 existing label(s)
0 missing value(s) (tokens with '-' label)
Existing: 'LOC', 'MISC', 'ORG', 'PERSON'
✔ Good amount of examples for all labels
✔ Examples without occurrences available for all labels
✔ No entities consisting of or starting/ending with whitespace
✔ No entities consisting of or starting/ending with punctuation
```

(All labels show up as existing because of the v3 initialization step.)

**Reply by m-nlp-q** [reply]
Thank you @adrianeboyd , I spent a number of days trying to figure out how to run the training.
I reproduced the training and I have two points:

1. I checked that the config file created by `python -m spacy init config -p ner -l el config.cfg` is different from the config file that has the `el_core_news_lg` that I used to train the model, so `spacy-legacy` is not necessary. 
2. Converting the data this way does not use the Greek model pipeline as a base. Does this have an impact on the `.spacy` data?

**Reply by adrianeboyd** [reply]
Using the sentencizer vs. the parser from `el_core_news_sm` only affects the position of the artificial document boundaries. Unless you have entities with sentence-final punctuation in the middle, it shouldn't really matter.

The document boundaries might be slightly different with `el_core_news_sm`, but the main concern here is to create shorter training documents that you can shuffle. The language `el` vs. `xx` doesn't matter for the sentencizer.

**Reply by m-nlp-q** [reply]
I see @adrianeboyd . You have been really helpful!
